Interpretation translation

AdritRao opened the pull request:
<!--

This source file is part of the Stanford Biodesign for Digital Health open-source project

SPDX-FileCopyrightText: 2022 Stanford University and the project authors (see CONTRIBUTORS.md)

SPDX-License-Identifier: MIT

-->

# Interpretation and Summary Translation

## :recycle: Current situation & Problem
Currently, all interpretations (single and multiple resources) as well as resource summaries are provided in English. Enabling multilingual support will allow users to communicate with the LLM in their preferred language.

## :bulb: Proposed solution
The prompt is modified based on the user's preferred language on their iPhone. The summary and initial multiple/single resource interpretation are provided in the preferred language. The interpretation prompt is also modified so that the LLM will reply to the user in the same language the user chats in.

## :gear: Release Notes 
The main code addition was done in the [Prompt.swift](https://github.com/StanfordBDHG/LLMonFHIR/blob/interpretation_translation/LLMonFHIR/Settings/Prompt.swift) as follows:

```
    var prompt: String {
        let preferredLanguage = Locale.preferredLanguages[0]
        var prompt = UserDefaults.standard.string(forKey: rawValue) ?? defaultPrompt
        prompt += "\n Chat with the user in \(preferredLanguage)"
        return prompt
    }
```
This tells the LLM to chat with the user in the iPhone system preferred language. The LLM interpretation prompt for single/multiple resources was also modified in [Localizable.strings](https://github.com/StanfordBDHG/LLMonFHIR/blob/interpretation_translation/LLMonFHIR/Resources/Localizable.strings) so that the LLM chats with the user in the same language the user chats in.

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

PSchmiedmayer commented on LLMonFHIR.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:
Seems like a few of the Swift Packages have been reset to some older versions, this might also be why the builds are failing. Be sure that you update your Swift Packages to the latest version using the File > Swift Packages menu in Xcode.

PSchmiedmayer commented on LLMonFHIR/Settings/Prompt.swift:
```diff
@@ -45,7 +45,10 @@ enum Prompt: String {
-        UserDefaults.standard.string(forKey: rawValue) ?? defaultPrompt
+        let preferredLanguage = Locale.preferredLanguages[0]
+        var prompt = UserDefaults.standard.string(forKey: rawValue) ?? defaultPrompt
+        prompt += "\n Chat with the user in \(preferredLanguage)"
```
We should ideally also localize this string, just to be consistent with the rest of the application, we also localize all prompts.

PSchmiedmayer commented on LLMonFHIR/Settings/Prompt.swift:
```diff
@@ -45,7 +45,10 @@ enum Prompt: String {
     
     var prompt: String {
-        UserDefaults.standard.string(forKey: rawValue) ?? defaultPrompt
+        let preferredLanguage = Locale.preferredLanguages[0]
```
You can probably directly move this in the string interpolation at line 50 as you don't really use this property somewhere else.

A more general comment: I would generally suggest to keep creating properties closer to their first use, e.g. in this case right above line 50 if you want to keep it a constant property. Makes the code easier to read as this is less of a context switch for the reader.

PSchmiedmayer reviewed (changes requested):
Thank you for the improvement @AdritRao! Seems to work very well, and the code diff is very small, which is great to see!

We discussed to create a PR with the current state, but looking at the current state and the small diff, I would suggest to add localization for 1-2 languages we speak.
I am happy to take a look at a German (and possibly French translation, even though that might be very rusty); maybe you can cover another language that you speak so we at least review the output of an LLM that does this for plausibility.
I think that would just round up the PR and might make even nicer screenshots to demonstrate a full localization with the buttons and texts also being localized. You can easily start this by using the localization functionality in Xcode and adding a new language, e.g., German:
<img width="1024" alt="Screenshot 2023-07-18 at 9 39 43 PM" src="https://github.com/StanfordBDHG/LLMonFHIR/assets/28656495/b40f4a3c-3845-433e-be43-c350dff60c52">
<img width="1024" alt="Screenshot 2023-07-18 at 9 40 05 PM" src="https://github.com/StanfordBDHG/LLMonFHIR/assets/28656495/cf212930-133c-4329-8867-ce9144ada5ef">

It would also be great if you could update the screenshots in the README to reflect the new updates, e.g., the chat with all resources feature and a screenshot that is localized. I just created the initial ones using a red background of the Desktop and the simulator and cropping them all in the same format; you might find a smarter way to do this.

AdritRao commented:
Hi @PSchmiedmayer, thank you for your feedback on the PR! I have updated the code based on your comments as well as the README with new screenshots featuring the "Chat with all health records" and translation feature. Regarding the translation capabilities, I have added support via localizable files (generated using GPT) for Spanish, Chinese, German, and French.

PSchmiedmayer commented on README.md:
```diff
@@ -17,10 +17,13 @@ SPDX-License-Identifier: MIT
+ <img alt="AllHealthRecords" src="Figures/LLMonFHIR_Screenshots_Main.png">
+</p>
+
+LLMonFHIR has multilingual support. The LLM is prompt engineered to converse with the user based on their system language. The full app currently translates to English, Spanish, Chinese, German, and French.
```
I would suggest to swap this paragraph and the next one. I think that it would look nicer if the "This repository contains the ..." paragraph is the first one below the first heading, the sentence below is then below the second figure.

I would also suggest to rephrase the paragraph a bit:
```suggestion
LLMonFHIR supports multiple languages. The LLM is prompt-engineered to converse with users based on their system language. The application is currently translated into English, Spanish, Chinese, German, and French.
```

PSchmiedmayer reviewed (changes requested):
Thank you for the improvements and also taking the time to greatly improve the README @AdritRao!

We can merge the PR once the few smaller comments have been addressed, looking forward to the new functionality in LLM on FHIR! 🚀

AdritRao commented:
Hi @PSchmiedmayer, thank you for your comments on the PR! I finished addressing the comments and added a localization for the translation prompt which is added to the main interpretation/summary prompts. I also updated the README file. Thanks!

PSchmiedmayer commented on LLMonFHIR/Settings/Prompt.swift:
```diff
@@ -46,7 +44,9 @@ enum Prompt: String {
     var prompt: String {
         var prompt = UserDefaults.standard.string(forKey: rawValue) ?? defaultPrompt
-        prompt += Locale.preferredLanguages[0]
+        var promptTranslation = String(localized: "FHIR_PROMPT_TRANSLATION \("%@")")
+        promptTranslation = promptTranslation.replacingOccurrences(of: Prompt.promptPlaceholder, with: Locale.preferredLanguages[0])
+        prompt += promptTranslation
```
You can probably compress the three lines down to the following statement:
```suggestion
        prompt += String(localized: "FHIR_PROMPT_TRANSLATION \(Locale.preferredLanguages[0])") 
```
Compared to the use case with the prompt where we allow the user to adjust the prompt in the settings, we don't offer that customization with the translation statement and therefore can just use normal string interpolation for localized strings to enable the same functionality. J

ust to be sure, please test this in the application so this works as expected.

PSchmiedmayer commented on LLMonFHIR/Resources/de.lproj/Localizable.strings:
```diff
@@ -0,0 +1,170 @@
+Stellen Sie sich am Anfang nicht vor und beginnen Sie mit Ihrer Interpretation.
+Stellen Sie sicher, dass Ihre Antwort in derselben Sprache erfolgt, in der der Benutzer Ihnen schreibt.
+Die Zeitform sollte in der Gegenwart sein.
+Kommunizieren Sie in derselben Sprache wie der Benutzer.
```
You can remove the `Kommunizieren Sie in derselben Sprache wie der Benutzer.` line now that we append it outside of the prompt. Same for all other languages as well.

PSchmiedmayer commented on LLMonFHIR/Resources/en.lproj/Localizable.strings:
```diff
@@ -105,6 +108,7 @@ The following JSON representation defines the FHIR resource that you should inte
 Provide a short one-sentence summary of the resource in less than 20 words.
 Do NOT respond with more content than the single line containing the summary.
 Directly provide the content without any additional structure.
+Chat with the user in
```
You can remove the `Chat with the user in` line now that we append it outside of the prompt. Same for all other languages as well.

PSchmiedmayer commented on LLMonFHIR/Settings/Prompt.swift:
```diff
@@ -17,10 +17,8 @@ enum Prompt: String {
-    
     static let promptPlaceholder = "%@"
     
-    
```
We have a style guide in place that we have two empty lines between blocks of elements in a type, such as static properties and computed properties. It would be great if you can re-add the lines here.

PSchmiedmayer commented on LLMonFHIR/Settings/Prompt.swift:
```diff
@@ -17,10 +17,8 @@ enum Prompt: String {
     
     case interpretMultipleResources = "prompt.interpretMultipleResources"
     
-    
```
We have a style guide in place that we have two empty lines between blocks of elements in a type, such as case statements and static properties. It would be great if you can re-add the lines here.

PSchmiedmayer reviewed (changes requested):
Thank you for the changes, I had a few smaller comments to finish up the PR. Thank you for all the work that goes into these changes! 🚀

PSchmiedmayer reviewed (approved):
Thank you for the improvements @AdritRao! Feel free to merge the PR one the last comment is addressed by just reading the removed empty line here: https://github.com/StanfordBDHG/LLMonFHIR/pull/26/files#r1277077261.

Apart from this the PR looks great, a strong contribution to LLM on FHIR and we look forward to hear how users will benefit from this feature enabling them to make health records even more accessible! 🚀

vishnuravi commented:
Great job @AdritRao!

AdritRao commented:
Thank you @PSchmiedmayer and @vishnuravi! I removed the empty line and I will merge the PR! I am looking forward to making more contributions to LLMonFHIR.
